plugin/plugin.go

```diff
@@ -382,23 +406,151 @@ func (p *AzurePlugin) ListHosts(ctx context.Context, req *pb.ListHostsRequest) (
+						}
+						var netInfo networkInfo
+						for _, ipconf := range *iface.InterfacePropertiesFormat.IPConfigurations {
+							if ipconf.PrivateIPAddress != nil {
+								netInfo.IpAddresses = append(netInfo.IpAddresses, *ipconf.PrivateIPAddress)
+							}
+							if ipconf.PublicIPAddress != nil {
+								if ipconf.PublicIPAddress.ID == nil {
+									return nil, fmt.Errorf("ip configuration %q has public IP address info but nil id", *ipconf.ID)
+								}
+								//ipResGroup, ipName, err := splitId(*ipconf.PublicIPAddress.ID, constMsNetworkService, constPublicIpAddressesResource)
+								if err != nil {
+									return nil, fmt.Errorf("error splitting public ip address id %q: %w", *ipconf.PublicIPAddress.ID, err)
+								}
+								pipName := extractResourceSuffix(*ipconf.PublicIPAddress.ID)
+								pipInfo, err := pipClient.GetVirtualMachineScaleSetPublicIPAddress(ctx, resourceGroup, *vmss.Name, *vmssvm.InstanceID, ifName, *ipconf.Name, pipName, "")
+								if err != nil {
+									if err != nil {
+										return nil, fmt.Errorf("error fetching public IP information with resource group %q and name %q: %w", resourceGroup, pipName, err)
+									}
+								}
+								if pipInfo.PublicIPAddressPropertiesFormat == nil {
+									return nil, fmt.Errorf("nil public ip address properties format for public ip %q", *ipconf.PublicIPAddress.ID)
+								}
+								if pipInfo.PublicIPAddressPropertiesFormat.IPAddress != nil {
+									netInfo.IpAddresses = append(netInfo.IpAddresses, *pipInfo.PublicIPAddressPropertiesFormat.IPAddress)
+								}
+							}
+						}
+						vmToNetworkMap[*vmssvm.ID] = netInfo
+					}
+
+				}
 			}
-			vmToNetworkMap[vmId] = netInfo
 		}
 	}
 
 	ret := &pb.ListHostsResponse{}
 	for k, v := range vmToNetworkMap {
-		_, externalName, err := splitId(k, constMsComputeService, constVirtualMachinesResource)
-		if err != nil {
-			externalName = ""
+		var host *pb.ListHostsResponseHost
+		// see if we're dealing with a VMSS instance or a VM
+		if strings.Contains(k, constMsComputeService+"/"+constVirtualMachineScaleSetsResource) {
+			externalName, err := getExternalNameforVMSSInstance(k)
+			if err != nil {
+				externalName = ""
+			}
+			setId := getSetForVMSSInstance(k)
+			host = &pb.ListHostsResponseHost{
+				ExternalId:   k,
+				ExternalName: externalName,
+				IpAddresses:  v.IpAddresses,
+				SetIds:       resourceToSetMap[setId],
+			}
 		}
-
-		host := &pb.ListHostsResponseHost{
-			ExternalId:   k,
-			ExternalName: externalName,
-			IpAddresses:  v.IpAddresses,
-			SetIds:       resourceToSetMap[k],
+		if strings.Contains(k, constMsComputeService+"/"+constVirtualMachinesResource) {
+			_, externalName, err := splitId(k, constMsComputeService, constVirtualMachinesResource)
+			if err != nil {
+				externalName = ""
+			}
+			host = &pb.ListHostsResponseHost{
+				ExternalId:   k,
+				ExternalName: externalName,
+				IpAddresses:  v.IpAddresses,
+				SetIds:       resourceToSetMap[k],
+			}
 		}
```

Comment:
Looks like these statements are exclusive, so maybe use a switch statement to make that clear?